sdk/core/Azure.Core/src/RehydrationToken.Serialization.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ClientModel.Primitives;
+using System.Text.Json;
+
+#nullable enable
+
+namespace Azure.Core
+{
+    public partial struct RehydrationToken : IJsonModel<RehydrationToken>
+    {
+        internal RehydrationToken DeserializeRehydrationToken(JsonElement element, ModelReaderWriterOptions options)
+        {
+            if (element.ValueKind == JsonValueKind.Null)
+            {
+                return default;
+            }
+            string? id = null;
+            string version = string.Empty;
+            string headerSource = string.Empty;
+            string nextRequestUri = string.Empty;
+            string initialUri = string.Empty;
+            RequestMethod requestMethod = default;
+            string? lastKnownLocation = null;
+            string finalStateVia = string.Empty;
+
+            foreach (var property in element.EnumerateObject())
+            {
+                if (property.NameEquals("id"u8))
+                {
+                    if (property.Value.ValueKind == JsonValueKind.Null)
+                    {
+                        continue;
+                    }
+                    id = property.Value.GetString()!;
+                }
+                if (property.NameEquals("version"u8))
+                {
+                    version = property.Value.GetString()!;
+                    continue;
+                }
+                if (property.NameEquals("headerSource"u8))
+                {
+                    headerSource = property.Value.GetString()!;
+                    continue;
+                }
+                if (property.NameEquals("nextRequestUri"u8))
+                {
+                    nextRequestUri = property.Value.GetString()!;
+                    continue;
+                }
+                if (property.NameEquals("initialUri"u8))
+                {
+                    initialUri = property.Value.GetString()!;
+                    continue;
+                }
+                if (property.NameEquals("requestMethod"u8))
+                {
+                    requestMethod = new RequestMethod(property.Value.GetString()!);
```

Comment:
We are using `!` here to assert that we always get a string value from the JSON -- what happens if we don't?  Are we confident that the error-experience is correct?  Will it throw a serialization exception or a NRE?  Which is correct in this case?

Reply:
Only `Id` and `LastKnowLocation` are nullable, the others are required. So, we should throw on null for them.
`RehydrationToken` is mainly used by us internally now, as you can see most of the properties are internal. If any of the required fields are null, there is something wrong within our implementation.

Reply:
We should either Debug.Assert that they are not null or throw an exception if they are.

Reply:
Makes sense, adding `Debug.Assert`.